include/zephyr/drivers/rtc.h

```diff
@@ -0,0 +1,345 @@
+
+/**
+ * @brief RTC alarm configuration
+ *
+ * @param compare Counter compare value at which the handler will be invoked
+ * @param handler Handler invoked once the alarm is triggered
+ * @param user_data Optional user data passed to the handler when invoked
+ *
+ * @note This handler is likely to be invoked from within an ISR
+ * context
+ *
+ * @note The handler and user data may be lost when system resets
+ */
+struct rtc_alarm_config {
+	uint32_t compare;
+	rtc_alarm_triggered_handler_t handler;
+	void *user_data;
+};
+
+/**
+ * @typedef rtc_counter_set
+ * @brief API for setting RTC counter value
+ */
+typedef int (*rtc_counter_set)(const struct device *dev, uint32_t count);
+
+/**
+ * @typedef rtc_counter_get
+ * @brief API for getting RTC counter value
+ */
+typedef int (*rtc_counter_get)(const struct device *dev, uint32_t *count);
+
+/**
+ * @typedef rtc_alarm_config_set
+ * @brief API for setting RTC alarm configuration
+ */
+typedef int (*rtc_alarm_config_set)(const struct device *dev, size_t id,
+				    const struct rtc_alarm_config *config);
+
+/**
+ * @typedef rtc_alarm_config_get
+ * @brief API for getting RTC alarm configuration
+ */
+typedef int (*rtc_alarm_config_get)(const struct device *dev, size_t id,
+				    struct rtc_alarm_config *config);
+
+/**
+ * @typedef rtc_alarm_enable
+ * @brief API for enabling RTC alarm
+ */
+typedef int (*rtc_alarm_enable)(const struct device *dev, size_t id);
+
+/**
+ * @typedef rtc_alarm_disable
+ * @brief API for disabling RTC alarm
+ */
+typedef int (*rtc_alarm_disable)(const struct device *dev, size_t id);
+
+/**
+ * @typedef rtc_alarm_triggered_test
+ * @brief API for testing if RTC alarm triggered event has occurred
+ */
+typedef int (*rtc_alarm_triggered_test)(const struct device *dev, size_t id);
+
+/**
+ * @typedef rtc_alarm_triggered_clear
+ * @brief API for clearing RTC alarm triggered event
+ */
+typedef int (*rtc_alarm_triggered_clear)(const struct device *dev, size_t id);
+
+/**
+ * @brief RTC driver API
+ */
+__subsystem struct rtc_driver_api {
+	rtc_counter_set_t counter_set;
+	rtc_counter_get_t counter_get;
+	rtc_alarm_config_set_t alarm_config_set;
+	rtc_alarm_config_get_t alarm_config_get;
+	rtc_alarm_enable_t alarm_enable;
+	rtc_alarm_disable_t alarm_disable;
+	rtc_alarm_triggered_test_t alarm_triggered_test;
+	rtc_alarm_triggered_clear_t alarm_triggered_clear;
+};
+
+/**
+ * @brief API for setting RTC counter value
+ *
+ * @param dev Device instance
+ * @param count Counter value to set
+ *
+ * @return 0 if successful
+ * @return -ENOSYS if API is not implemented
+ * @return -errno code if failure
+ */
+__syscall int rtc_counter_set(const struct device *dev, uint32_t count);
+
+static inline int z_impl_rtc_counter_set(const struct device *dev, uint32_t count)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	if (api->counter_set == NULL) {
+		return -ENOSYS;
```

Comment:
It seems like all of the API calls are optional. Is that intentional?

If any API calls are optional, it should be explicitly called out (but there should be a good reason for that).